pkg/services/ngalert/schedule/schedule_unit_test.go

```diff
@@ -644,45 +693,64 @@ func TestSchedule_ruleRoutine(t *testing.T) {
 			updateChan <- ruleVersionAndPauseStatus{ruleFp + 1, false}
 
 			require.Eventually(t, func() bool {
-				return len(sender.Calls) > 0
+				return len(sender.Calls()) > 0
 			}, 5*time.Second, 100*time.Millisecond)
 
 			require.Empty(t, sch.stateManager.GetStatesForRuleUID(rule.OrgID, rule.UID))
 			sender.AssertNumberOfCalls(t, "Send", 1)
-			args, ok := sender.Calls[0].Arguments[2].(definitions.PostableAlerts)
-			require.Truef(t, ok, fmt.Sprintf("expected argument of function was supposed to be 'definitions.PostableAlerts' but got %T", sender.Calls[0].Arguments[2]))
+			args, ok := sender.Calls()[0].Arguments[2].(definitions.PostableAlerts)
+			require.Truef(t, ok, fmt.Sprintf("expected argument of function was supposed to be 'definitions.PostableAlerts' but got %T", sender.Calls()[0].Arguments[2]))
 			require.Len(t, args.PostableAlerts, expectedToBeSent)
 		})
 	})
 
 	t.Run("when evaluation fails", func(t *testing.T) {
-		rule := models.AlertRuleGen(withQueryForState(t, eval.Error))()
-		rule.ExecErrState = models.ErrorErrState
+		t.Parallel()
```

Comment:
I am on the fence about making tests parallel. How much time does it save? On my machine it saves 500ms. There is only one test that contributes 90% of time.

Reply:
The time it saves at runtime depends on the tests. Here I tried to prioritize parallelizing those which were the slowest. Seeing a 500ms saving in a single test says a lot, but we should expect the time saving for a single test parallelized to be negligible. The true time saving should be seen when the vast majority of tests in the repo are being parallelized, since it's actually thousands of them running in different goroutines, and that's where Go shines. Most of them are really fast, but not collectively fast if they run sequentially, which is the case as of today.
We should aim to run every test with `t.Parallel()`, even if trivial, and only not use it if there is an extraordinary reason not to (e.g. we really need to call an OS primitive or access a global symbol from the standard library). The main reason behind that is not only testing time and better utilization of resources, but understanding that we're also testing how we thought about concurrency while writing our code (both runtime and tests).
Put it another way: if we write a test that doesn't use `t.Parallel()` we should then be prepared to answer: "why am I writing code that cannot be run in parallel?".

Reply:
Go's test framework does support the execution of tests in parallel across different packages. I am even ok with running top level suites (test functions) in parallel because they usually isolated.  However, I think running all tests in parallel complicates the set-up of the tests, and therefore affects readability and maintainability. In this particular case, I do not see any significant benefit in doing it. If problem is that tests are slow, we need to improve its performance. I can do that.

Reply:
That's true, Go makes a binary for each package and runs them in parallel, and `t.Parallel()` only affects tests run within each binary. What I meant to say was exactly the opposite of calling `t.Parallel()` just to have tests run faster, but to change the mindset around how we think about testing. But that was just an opinion, and it has nothing to do with the objective quality of any part of the code.